File: Authentication/src/main/java/com/rectruit/authentication/jwt/JwtAuthenticationFilter.java
package com.rectruit.authentication.jwt;

import com.rectruit.authentication.service.UserDetailsServiceImpl;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.MalformedJwtException;
import io.jsonwebtoken.UnsupportedJwtException;
import jakarta.servlet.*;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.annotation.Order;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

@Component
@Order(2)
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    @Autowired
    private JwtUtils jwtUtils;

    @Autowired
    private UserDetailsServiceImpl userDetailsService;

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
            throws ServletException, IOException {
        try {
            String jwt = parseJwt(request);
            if (jwt != null && jwtUtils.validateJwtToken(jwt) && (boolean) request.getAttribute("ServiceAuthenticated")) {
                String id = jwtUtils.getIdFromJwtToken(jwt);

                UserDetails userDetails = userDetailsService.loadUserById(id);
                UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(
                        userDetails, "", userDetails.getAuthorities());
                authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));

                SecurityContextHolder.getContext().setAuthentication(authentication);
            }
        } catch (
                MalformedJwtException e) {
            //System.out.println("Invalid JWT token: " + e.getMessage());
            response.sendError(401, "Invalid JWT token: " + e.getMessage());
        } catch (
                ExpiredJwtException e) {
            //System.out.println("JWT token is expired: " + e.getMessage());
            response.sendError(401, "JWT token is expired: " + e.getMessage());
        } catch (
                UnsupportedJwtException e) {
            //System.out.println("JWT token is unsupported: " + e.getMessage());
            response.sendError(401, "JWT token is unsupported: " + e.getMessage());
        } catch (IllegalArgumentException e) {
            //System.out.println("JWT claims string is empty: " + e.getMessage());
            response.sendError(401, "JWT claims string is empty: " + e.getMessage());
        }
        catch (Exception e) {
            //System.out.println("Cannot set user authentication: " + e.getMessage());
            response.sendError(500, "Cannot set user authentication: " + e.getMessage());
        }

        chain.doFilter(request, response);
    }

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        String path = request.getServletPath();
        return path.startsWith("/auth/");
    }

    private String parseJwt(HttpServletRequest request) {
        String headerAuth = request.getHeader("Authorization");

        if (StringUtils.hasText(headerAuth) && headerAuth.startsWith("Bearer ")) {
            return headerAuth.substring(7);
        }

        return null;
    }

}

File: Gateway/src/main/kotlin/com/recruiter/gateway/jwt/JwtAuthenticationFilter.kt
package com.recruiter.gateway.jwt

import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import io.jsonwebtoken.UnsupportedJwtException
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.server.reactive.ServerHttpRequest
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatcher
import org.springframework.security.web.server.util.matcher.ServerWebExchangeMatchers
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import java.nio.charset.Charset

class JwtAuthenticationFilter(private val jwtSecret: String) : WebFilter {

    private fun parseJwt(request: ServerHttpRequest): String? {
        val headerAuth = request.headers.getFirst("Authorization")
        return if (StringUtils.hasText(headerAuth) && headerAuth!!.startsWith("Bearer ")) {
            headerAuth.substring(7)
        } else null
    }

    private fun getUserId(token: String?): String {
        return Jwts.parser().setSigningKey(jwtSecret).parseClaimsJws(token).body.subject
    }

    private fun validateJwtToken(authToken: String): Pair<Boolean, String> {
        return try {
            Jwts.parser().setSigningKey(jwtSecret).parseClaimsJws(authToken)
            Pair(true, "")
        } catch (e: MalformedJwtException) {
            Pair(false, "Invalid JWT token: " + e.message)
        } catch (e: ExpiredJwtException) {
            Pair(false, "JWT token is expired: " + e.message)
        } catch (e: UnsupportedJwtException) {
            Pair(false, "JWT token is unsupported: " + e.message)
        } catch (e: IllegalArgumentException) {
            Pair(false, "JWT claims string is empty: " + e.message)
        }
    }

    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val token = parseJwt(exchange.request)
        if (token != null) {
            val userId = getUserId(token)

            val validationPair = validateJwtToken(token)

            return if (validationPair.first) {
                val auth: Authentication = UsernamePasswordAuthenticationToken(userId, null, null)
                chain.filter(exchange).contextWrite(ReactiveSecurityContextHolder.withAuthentication(auth))
            } else {
                unauthorized(exchange, validationPair.second)
            }
        }
        return unauthorized(exchange)
    }

    private fun unauthorized(
        exchange: ServerWebExchange,
        errorMessage: String = "Unauthorized Access. Please provide a valid token."
    ): Mono<Void> {
        exchange.response.statusCode = HttpStatus.UNAUTHORIZED
        exchange.response.headers.set(HttpHeaders.WWW_AUTHENTICATE, "Bearer realm=\"Unauthorized\", charset=\"UTF-8\"")

        val buffer = exchange.response.bufferFactory().wrap(errorMessage.toByteArray(Charset.defaultCharset()))
        return exchange.response.writeWith(Mono.just(buffer))
    }
}